transport-native-io_uring/src/main/java/io/netty/channel/uring/IOUringCompletionQueue.java

```diff
@@ -51,37 +54,36 @@
     this.ringSize = ringSize;
     this.ringAddress = ringAddress;
     this.ringFd = ringFd;
+
+    this.ringMask = PlatformDependent.getInt(kringMaskAddress);
+    this.ringHead = PlatformDependent.getInt(kHeadAddress);
+  }
+
+  public boolean hasCompletions() {
+      return ringHead != PlatformDependent.getIntVolatile(kTailAddress);
   }
 
   public int process(IOUringCompletionQueueCallback callback) {
+      int tail = PlatformDependent.getIntVolatile(kTailAddress);
       int i = 0;
-      for (;;) {
-          long head = toUnsignedLong(PlatformDependent.getIntVolatile(kHeadAddress));
-          if (head != toUnsignedLong(PlatformDependent.getInt(kTailAddress))) {
-              long index = head & toUnsignedLong(PlatformDependent.getInt(kringMaskAddress));
-              long cqe = index * CQE_SIZE + completionQueueArrayAddress;
-
-              long udata = PlatformDependent.getLong(cqe + CQE_USER_DATA_FIELD);
-              int res = PlatformDependent.getInt(cqe + CQE_RES_FIELD);
-              long flags = toUnsignedLong(PlatformDependent.getInt(cqe + CQE_FLAGS_FIELD));
-
-              //Ensure that the kernel only sees the new value of the head index after the CQEs have been read.
-              PlatformDependent.putIntOrdered(kHeadAddress, (int) (head + 1));
-
-              int fd = (int) (udata >> 32);
-              int opMask = (int) udata;
-              short op = (short) (opMask >> 16);
-              short mask = (short) opMask;
-
-              i++;
-              if (!callback.handle(fd, res, flags, op, mask)) {
-                  break;
-              }
-          } else {
-              if (i == 0) {
-                  return -1;
```

Comment:
This is not present in the new code?

Reply:
I think it was redundant. The method returns count of completions processed, so didn't see a reason for converting 0 to -1, maybe @1Jo1 can confirm?

Reply:
@njhill yeah it is redundant, it looks fine to me, just replace -1 to 0 :)